### ISSUE TITLE
Signing an .appx file corrupts the .appx

### ISSUE THREAD

**Issue by st-gr** [mtrojnar/osslsigncode#367]
Hi,

First of all, I would like to thank you all for creating this amazing tool that eventually will allow me to build and sign Windows apps running in a Linux/wine docker container.

Irrespective of me building osslsigncode from source or downloading and using the latest release (2.7) downloaded from GitHub the resulting appx gets corrupted.

The appx is created using apache cordova version 11 (yes, I still have to build Windows 8 UWP apps ;-)):
```
cordova create testapp
cordova platforms add windows
cordova build --debug --arch=x64
```
_(I prevented MSBuild from signing by replacing signtool.exe with a mock executable that returns error code 0)._

Here is the command I used to sign the .appx on Windows 10 and Ubuntu 22.04/wine 8.0.2 stable:
```
osslsigncode sign -in CordovaApp.Windows10_1.0.0.0_x64_debug.appx -out osslsigncode-signed.appx -pkcs12 ..\..\CordovaApp_TemporaryKey.pfx
Hash method is SHA256
Warning: Ignore -h option, use the hash algorithm specified in AppxBlockMap.xml
Signing as a package
Succeeded
```
BTW: The release build 2.7 ran under wine, but the latest source build returned errors and did not sign the file:
```
Unable to load provider: legacy
Warning: Legacy mode disabled
```

As a comparison, I signed the .appx using signtool:

```
PS T:\appxsign-test\platforms\windows\AppPackages\CordovaApp.Windows10_1.0.0.0_x64_debug_Test> ."C:\Program Files (x86)\Windows Kits\10\App Certification Kit\signtool.exe" sign /fd sha256 /f ..\..\CordovaApp_TemporaryKey.pfx signtool-signed.appx
Done Adding Additional Store
Successfully signed: signtool-signed.appx
```

When I inspect the corrupted .appx with Z-zip I don't see the packaged files, only one file shows.
I then tried to repair the .appx using WinZip which didn't help as this compromised the signature.
![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/838dec34-9409-4d2d-be8d-a8cda2624503)

I uploaded the unsigned .appx, the .pfx, and the two signed .appx in this .zip file:
[appx-files.zip](https://github.com/mtrojnar/osslsigncode/files/14446481/appx-files.zip)

I haven't checked with a hex editor, yet. There might be a bug in the packaging of the .appx.
Microsoft provides an MSIX packaging tool (src/makemsix) that also creates .appx. I thought about looking there for inspiration [https://github.com/microsoft/msix-packaging](url)

Your tool is the last puzzle piece for me to be able to build and sign my apps in a Linux/wine container. It took me a while to get MSBuild to behave..

Maybe I am just using your tool with the wrong key format or parameters?

Please advise.
st-gr

**Comment by mtrojnar** [comment]
> Irrespective of me building osslsigncode from source or downloading and using the latest release (2.7)

What exactly do you mean by "source" here: the current GitHub master branch or something else?

**Comment by st-gr** [comment]
Hi mtrojnar,

Sorry for not being specific enough. Yes, I built (VS Buildtools 2022) and tested the current master branch. I didn't know how to create a static build and tested the VS build with the separate libraries on wine which failed. Need to look for static build instructions next time.

Meanwhile I made some observations using the HxD and 010 Hex editors.
I created a minimal .appx with 10 KB size. You can download it here:
[appx-files.zip](https://github.com/mtrojnar/osslsigncode/files/14459256/appx-files.zip)

**1. Signtool keeps the .zip file mostly untouched (files are in the same order after signing compared to osslsigncode):**

![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/8a5cf0f6-47bd-405f-b3a4-e93990527635)

Signtool output vs. osslsigncode:
![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/e2ae38b0-2e86-4448-9537-e54e8e7524c8)

**2. Osslsigncode seems to add 8 bytes (64 bit) to the zip data stream:**
![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/10bd69ba-92e8-433e-96f1-64c2527cd50b)

![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/251f3a23-7aa9-4b9b-bc1f-dde9c9ec545c)

![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/204565f3-909f-4526-8623-47ac6259d088)

**3. Osslsigncode is missing the uncompressed and compressed size compared to a WinRAR fixed file:**
![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/fe86ce2d-8be4-4f41-97f6-e6cdf8842c7a)

**4. All but one file offset are wrong. The correct file offset originates from the added AppxSignature.p7x**
I manually fixed them in the hex editor:
![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/b30241b7-7ac9-45fd-a9ec-f1ccb6dd13cc)

**5. The manually adjusted file appx\rebuilt.osslsigncode-signed-edited.appx passed WinRAR tests, shows that it has a digital signature and loads via installer:**
![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/8b7660e8-7ff4-454d-971c-b9f321bd4af2)

![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/9293d997-a4f9-4476-9dbc-6f93db2c581e)

So a manual fix would work and I established that osslsigncode is corrupting the appx zip archive.

The next step will be for me to look at your codebase where the zip (appx) is generated.

If I can't patch your code (because your C++ Kung Fu is better than mine ;-) then a post processing step that applies my manual fixes might do the trick.

Best,
st-gr

**Comment by st-gr** [comment]
Hi olzomal,

MakeAppx.exe is executed by MSBuild when cordova is calling MSBuild to build the .appx.
The trouble is that MakeAppx.exe refuses to run under wine (stable 8.0.2) as wine does not implement certain ntdll avl table functions (ReactOS does):

```
  Task Parameter:Output=C:\windows\temp\hello\platforms\windows\AppPackages\CordovaApp.Windows10_1.0.0.0_x64_debug_Test\CordovaApp.Windows10_1.0.0.0_x64_debug.appx (TaskId:140)
  C:\Program Files (x86)\Windows Kits\10\bin\10.0.17763.0\x64\MakeAppx.exe pack /l /h sha256 /f build\windows\bld\package.map.txt /o /p C:\windows\temp\hello\platforms\windows\AppPackages\CordovaApp.Windows10_1.0.0.0_x
  64_debug_Test\CordovaApp.Windows10_1.0.0.0_x64_debug.appx   (TaskId:140)
wine: Call from 000000017002E8F8 to unimplemented function ntdll.dll.RtlLookupElementGenericTableAvl, aborting
wine: Unimplemented function ntdll.dll.RtlLookupElementGenericTableAvl called at address 000000017002E8F8 (thread 0500), starting debugger...
  Unhandled exception: unimplemented function ntdll.dll.RtlLookupElementGenericTableAvl called in 64-bit code (0x0000017002e8f8). (TaskId:140)
```

So I make use of the Microsoft [MSIX packager](https://github.com/microsoft/msix-packaging)  instead assuming that the Microsoft tool produces valid appx files.
I coded a small stub for MakeAppx.exe that accepts the /f parameter, creates a temporary folder of the files listed in the package.map.txt file, and then runs makemsix.exe using said folder.
That is how I got around the above MakeAppx.exe wine issue during the build. 

As osslsigntool is targeting to replace signtool to run under Linux (wine) I figure it should support signing any valid appx package I throw its way.
Signtool can sign my above makemsix packaged appx, so should osslsigntool.

I followed your suggestion and used MakeAppx.exe on a Windows 10 box at first look the MakeAppx bundled appx does not differ much from the makemsix one:
![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/0abc571b-f229-4579-8368-431d6b8c27dc)

The directory size differs dramatically though:
![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/399f75c7-4566-477f-b640-4f03183d564b)

The 7-zip properties dialog shows “Unsorted_CD” for my makemsix bundled file. It also shows that the makemsix bundled file seems to be compressed more efficiently. 
![image](https://github.com/mtrojnar/osslsigncode/assets/38470677/85e77c21-f239-4045-9f6c-36ebf1a1bc10)

Maybe that is why the Microsoft engineers chose this Zip64 format for their MSIX packager?
ChatGPT:
_"Unsorted_CD" in the context of 7-Zip's properties dialog likely refers to an "Unsorted Central Directory." In a ZIP file, the Central Directory (CD) is a section that contains records of all the files stored in the ZIP archive. Each record in the Central Directory provides information about a file, such as its name, size, compression method, and other metadata.

Normally, the records in the Central Directory are sorted in some way, often by file name or by the order in which the files were added to the archive. However, in some cases, the records in the Central Directory might not be sorted. This could happen for various reasons, such as the way the ZIP file was created or modified.

When 7-Zip displays "Unsorted_CD" in the properties dialog, it's indicating that the Central Directory records in the ZIP file are not sorted in the usual manner. This is generally just informational and shouldn't affect your ability to use or extract files from the ZIP archive._

However, there are differences and after I signed the MakeAppx bundled appx with osslsigncode 2.7 it worked!
Too bad I can't get MakeAppx.exe to run under Wine.

Now MSIX is the successor of Appx and other application bundle formats. MSIX is less restrictive compared to Appx. Possibly that is the reason why osslsigncode makes assumptions about the internal structure of the appx that are not present in an MSIX bundled appx? Remember: Signtool happily signed my MSIX packaged Appx.

**I wonder did anyone here got MakeAppx.exe to run under Wine stable 8.0.2 or are you using a different packager that works with osslsigntool?**


Some of the differences in the MSIX packaged Appx throw off osslsigncode, but not signtool. Possibly the unsorted central directory?

Strictly speaking, this is not your problem and I would understand that you won’t support this feature to support signing MSIX files. I might therefore revert to my envisioned post-processing step or ‘fix’ the MSIX packager to create a zip that is like the MakeAppx.exe generated ones.

Best,
st-gr

**Comment by st-gr** [comment]
Hi,

I discovered that fixing the sort order of the central directory enables **osslsigncode** to successfully sign **makemsix** packaged appx files. 

I've submitted a pull request to [microsoft/msix-packaging](https://github.com/microsoft/msix-packaging/pull/622) to address this by sorting the central directory in the same sequence as the local fileheader streams were added. This workaround resolves the reported issue for me. 

Feel free to close this issue. However, it's worth noting that while signtool could sign the appx with unsorted central directory entries, osslsigncode could not without corrupting the archive.

Best,
st-gr
